### PR TITLE
X11: Add OverrideRedirect attribute to new windows

### PR DESCRIPTION
**Pull request by stevebob**
This flag allows windows to avoid capture by tiling window managers such
as i3. Following this change, windows created with a minimum and maximum
size will not be tiled by the WM when first created. Note that the logic
for specifying the size hints of the window to prevent scaling was
already present. This change just adds the OverrideRedirect flag which
allow them to take effect.

I've tested this change on my laptop with archlinux + i3. i3 is a tiling window manager.

Prior to this change, regardless of a window's size, it would be tiled by the window manager. If I removed the window from tiling, it would take on its specified size.

Following this change, the behaviour above is still exhibited **unless** the window has a max size **and** min size specified, in which case it the window appears with its specified size, and isn't managed by the tiling window manager unless explicitly added.

Relevant docs:
Override Redirect Flag: https://tronche.com/gui/x/xlib/window/attributes/override-redirect.html
Window Attributes: https://tronche.com/gui/x/xlib/window/attributes
SDL2 source I used as a reference: https://github.com/SDL-mirror/SDL/blob/master/src/video/x11/SDL_x11window.c#L502

### CODE REVIEW

**Comment by Ralith** [comment]
[Authoritative documentation](https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html) does not obviously mention any special semantics regarding sizes being specified. What basis is there to expect the behavior of other window managers to match i3? Wouldn't it make more sense only to set the attribute when size limits are specified?

**Comment by stevebob** [comment]
Does the documentation say anything about the convention which window managers should follow regarding setting the size and position of windows? It would be neat if we could know for sure that if a window's size hints will be respected as long as the "override redirect" flag is set.

As far as I know there's no reason to set "override redirect" unless a min and max size are set. Shall I add a check to prevent the flag being set in the absence of a min and max size?

**Comment by Ralith** [comment]
I dug further and found some more pertinent documentation:

The [ICCM](https://tronche.com/gui/x/icccm/) makes it pretty clear that setting override-redirect is *not* appropriate to do by default, regardless of the user's size hints: "[In general, this should be done only if the pointer is grabbed while the window is mapped [...] The user will not be able to move, resize, restack, or transfer the input focus to override-redirect windows](https://tronche.com/gui/x/icccm/sec-4.html#s-4.1.10)."

The [EWMH](https://standards.freedesktop.org/wm-spec/wm-spec-latest.html) spec provides [some insight](https://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472629520) into the intended use of override-redirect: dropdown/popup menus, tooltips, notification bubbles, auto-completion suggestions, and drag-and-drop previews. It's how GUI toolkits implement those features without inadvertently getting window decorations on them. None of these are typical winit use-cases, although it might be interesting to explore supporting them (imagine drag-and-dropping animated 3D objects!).

> Does the documentation say anything about the convention which window managers should follow regarding setting the size and position of windows?

My understanding is that all size information specified by an X11 client is strictly advisory. Handling resizes well is better UX, even if in the unlikely event that the very best you can do is crop or black-bar; the user may not have space for the size your window ostensibly requires, or may have a bunch of dead space and no better use for it. X11 is designed in recognition of that.

The fact that tiling window managers ignore your size hints is, ultimately, a feature. As an i3 user myself, I don't want random applications to open in floating mode by default just because they think they know better--I can toggle floating myself, or set it to default based on WM_CLASS, if I want it. If it is absolutely necessary to float by default, e.g. if you're implementing a dialog box, the appropriate mechanism is to set `_NET_WM_WINDOW_TYPE` to `_NET_WM_WINDOW_TYPE_DIALOG`, but wanting stronger size hints is not, IMO, a sufficient justification to pretend to be a dialog box.

In conclusion, standards and conventions read to me as firmly contrary to use of override-redirect in this manner, so I am strongly against this change.

**Comment by stevebob** [comment]
Thanks for doing this research! I don't mind not merging this. I can just set windows to be dialogue boxes if I don't want them to be tiled.

Closing this PR.
